File: src/main/java/dev/zprestige/fire/module/combat/autotrap/TickListener.java
package dev.zprestige.fire.module.combat.autotrap;

import dev.zprestige.fire.Main;
import dev.zprestige.fire.event.bus.EventListener;
import dev.zprestige.fire.event.impl.TickEvent;
import dev.zprestige.fire.manager.playermanager.PlayerManager;
import dev.zprestige.fire.util.impl.BlockUtil;
import dev.zprestige.fire.util.impl.EntityUtil;
import net.minecraft.init.Blocks;
import net.minecraft.util.math.BlockPos;

import java.util.ArrayList;

public class TickListener extends EventListener<TickEvent, AutoTrap> {

    public TickListener(final AutoTrap autoTrap) {
        super(TickEvent.class, autoTrap);
    }

    @Override
    public void invoke(final Object object) {
        final PlayerManager.Player entityPlayer = EntityUtil.getClosestTarget(EntityUtil.TargetPriority.Range, module.placeRange.GetSlider() - 1.0f);
        if (entityPlayer == null || !BlockUtil.isPlayerSafe(entityPlayer) || module.multiTaskValid() || (module.disableOnSelfMove.GetSwitch() && !BlockUtil.getPosition().equals(module.pos))) {
            module.disableModule();
            return;
        }
        final BlockPos pos = entityPlayer.getPosition();
        final ArrayList<AutoTrap.Position> positions = module.getPossiblePlacePositions(pos);
        final int obsidian = Main.inventoryManager.getBlockFromHotbar(Blocks.OBSIDIAN);
        if (obsidian == -1) {
            module.disableModule();
        }
        switch (module.mode.GetCombo()) {
            case "Instant":
                int i = 0;
                for (final AutoTrap.Position position : positions) {
                    if (i > module.blocksPerTick.GetSlider() || Main.inventoryManager.getBlockFromHotbar(Blocks.OBSIDIAN) == -1 || (module.preventRotationRubberband.GetSwitch() && Main.rotationManager.maxRotations())) {
                        return;
                    }
                    Main.interactionManager.placeBlockWithSwitch(position.getPos(), module.rotate.GetSwitch(), module.packet.GetSwitch(), module.strict.GetSwitch(), obsidian);
                    module.addFade(position.getPos());
                    i++;
                }
                break;
            case "Tick":
                for (final AutoTrap.Position position : positions) {
                    if (module.preventRotationRubberband.GetSwitch() && Main.rotationManager.maxRotations()) {
                        return;
                    }
                    Main.interactionManager.placeBlockWithSwitch(position.getPos(), module.rotate.GetSwitch(), module.packet.GetSwitch(), module.strict.GetSwitch(), obsidian);
                    module.addFade(position.getPos());
                    return;
                }
                break;
        }
    }
}

File: src/main/java/dev/zprestige/fire/module/movement/anchor/TickListener.kt
package dev.zprestige.fire.module.movement.anchor

import dev.zprestige.fire.Main
import dev.zprestige.fire.event.bus.EventListener
import dev.zprestige.fire.event.impl.TickEvent
import dev.zprestige.fire.manager.playermanager.PlayerManager
import dev.zprestige.fire.util.impl.BlockUtil
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import java.util.*

class TickListener(anchor: Anchor) : EventListener<TickEvent, Anchor>(TickEvent::class.java, anchor) {

    override fun invoke(e: Any) {
        if (module.pos != null && mc.player.getDistanceSq(module.pos!!) > module.range.GetSlider() * 2.0f) {
            module.pos = null
        }
        if (module.pos != null && !Main.holeManager.contains(module.pos)) {
            module.pos = null
        }
        if (BlockUtil.isPlayerSafe(PlayerManager.Player(mc.player))) {
            module.pos = null
            module.timer.syncTime()
            return
        }
        if (module.pos == null) {
            val posses = TreeMap<Double, BlockPos>()
            for (holePos in Main.holeManager.holes) {
                if (holePos.isDouble) {
                    continue
                }
                val pos1 = holePos.pos
                val dist = mc.player.getDistanceSq(pos1)
                if (if (module.mode.GetCombo() == "Snap") mc.player.posY > pos1.getY() else mc.player.posY >= pos1.getY() && dist < module.range.GetSlider() * 2.0f) {
                    posses[dist] = pos1
                }
            }
            if (!posses.isEmpty()) {
                module.pos = posses.firstEntry().value
            }
        } else if (module.timer.getTime(1000)) {
            if (module.pos!!.getY() < mc.player.posY) {
                when (module.mode.GetCombo()) {
                    "Stop Motion" -> if (module.isOver(module.pos!!) && module.intersects(
                            module.pos!!
                        )
                    ) {
                        mc.player.motionX = 0.0
                        mc.player.motionZ = 0.0
                        module.setMovementsFalse()
                    }
                    "Pull" -> {
                        val bb = AxisAlignedBB(module.pos!!).shrink(0.35)
                        val speed = module.pullSpeed.GetSlider() / 10.0f
                        if (mc.player.posX > bb.maxX) {
                            mc.player.motionX = -speed.toDouble()
                        } else if (mc.player.posX < bb.minX) {
                            mc.player.motionX = speed.toDouble()
                        }
                        if (mc.player.posZ > bb.maxZ) {
                            mc.player.motionZ = -speed.toDouble()
                        } else if (mc.player.posZ < bb.minZ) {
                            mc.player.motionZ = speed.toDouble()
                        }
                        module.setMovementsFalse()
                    }
                    "Snap" -> {
                        if (!module.onGround.GetSwitch() || mc.player.onGround) {
                            mc.player.setPosition((module.pos!!.getX() + 0.5f).toDouble(),
                                mc.player.posY,
                                (module.pos!!.getZ() + 0.5f).toDouble()
                            )
                        }
                        mc.player.motionY = -module.snapFall.GetSlider().toDouble()
                        module.setMovementsFalse()
                    }
                }
            }
        }
    }
}